numpy/_core/src/multiarray/methods.c

```diff
@@ -888,28 +888,39 @@ array_finalizearray(PyArrayObject *self, PyObject *obj)
 }
 
 
+/*
+ * Default `__array_wrap__` implementation.
+ *
+ * If `self` is not a base class, we always create a new view, even if
+ * `return_scalar` is set. This way we preserve the (presumably important)
+ * subclass information.
+ * If the type is a base class array, we honor `return_scalar` and call
+ * PyArray_Return to convert any array with ndim=0 to scalar.
+ *
+ * By default, do not return a scalar (because this was always the default).
+ */
 static PyObject *
 array_wraparray(PyArrayObject *self, PyObject *args)
 {
     PyArrayObject *arr;
-    PyObject *obj;
+    PyObject *UNUSED = NULL;  /* for the context argument */
+    int return_scalar = 0;
 
-    if (PyTuple_Size(args) < 1) {
-        PyErr_SetString(PyExc_TypeError,
-                        "only accepts 1 argument");
-        return NULL;
-    }
-    obj = PyTuple_GET_ITEM(args, 0);
-    if (obj == NULL) {
+    if (!PyArg_ParseTuple(args, "O!|OO&:__array_wrap__",
+                &PyArray_Type, &arr, &UNUSED,
+                &PyArray_OptionalBoolConverter, &return_scalar)) {
         return NULL;
     }
-    if (!PyArray_Check(obj)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "can only be called with ndarray object");
-        return NULL;
+
+    if (return_scalar && Py_TYPE(self) == &PyArray_Type && PyArray_NDIM(arr) == 0) {
```

Comment:
the `NDIM` part can in principle be removed - `PyArray_Return` does that already.  Maybe slightly better? (If going that route, adjust the comment to `# Return scalar if NDIM==0.`

Reply:
Yeah, that is what I had... But then my test fails, because I somewhat prefer returning an `ndarray` rather than a subclass, and `PyArray_Return` won't do that.

Reply:
> Yeah, that is what I had... But then my test fails, because I somewhat prefer returning an `ndarray` rather than a subclass, and `PyArray_Return` won't do that.

OK, makes sense now!